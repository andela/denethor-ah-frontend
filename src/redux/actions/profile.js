import { REMOVE_OWN_PROFILE, SET_OWN_PROFILE, UPDATE_PROFILE, SET_NOTIFICATIONS } from './types';
import axios from '../../utils/axiosConfig';
import socket from '../../utils/socket';

const api = process.env.API_ROOT_URL;

export const getOwnProfile = id => async (dispatch) => {
  try {
    const { data: { data } } = await axios.get(`${api}/users/${id}/profile`);
 
    dispatch({
      type: SET_OWN_PROFILE,
      payload: data
    });

    socket.emit('addUserListener', data.id);

    socket.on('setNotifications', (notifications) => {
      dispatch({
        type: SET_NOTIFICATIONS,
        payload: notifications
      })
    });

    const userAverageRatingPromise = data.publishedArticles.reduce(async (accumulatorObject, article, index, { length }) => {
      const { totalArticlesWithRating, totalRating, ratingCount } = await accumulatorObject;

      const { data: { data, data: { length: count } } } = await axios.get(`${api}/articles/${article.id}/ratings`);

      const averageRating = data.reduce((average, { rating }) => average + rating/(count > 0 ? count : 1), 0);

      const articleHasRating = count > 0 ? 1 : 0;
      const articlesWithRating = totalArticlesWithRating + articleHasRating;
      const currentCount = ratingCount + count;
      const currentRating = Number(averageRating) + totalRating;

      return {
        ...((index === length - 1 && articlesWithRating > 0)
          && { averageRating: currentRating/articlesWithRating }),
        totalArticlesWithRating: articlesWithRating,
        totalRating: currentRating,
        ratingCount: currentCount
      }
    }, Promise.resolve({
      totalArticlesWithRating: 0,
      totalRating: 0,
      ratingCount: 0
    }));

    const userAverageRating = await Promise.resolve(userAverageRatingPromise);

    dispatch({
      type: UPDATE_PROFILE,
      payload: { userAverageRating }
    });

    return;
  } catch ({ response: { status } }) {
    switch (status) {
      case 404:
        return 'Not found.';

      case 500:
      case 502:
        return  'Server error.';

      default:
        return 'Unknown error.';
    }
  }
};

export const updateProfile = (id, updates) => async (dispatch) => {
  try {
    await axios.patch(`${api}/users/${id}/profile`, updates);

  dispatch({
    type: UPDATE_PROFILE,
    payload: updates
  });
  } catch ({ response: { status } }) {
    switch (status) {
      case 404:
        return 'Not found.';

      case 500:
      case 502:
        return  'Server error.';

      default:
        return 'Unknown error.';
    }
  }
};


export const uploadProfilePicture = (id, newProfilePicture) => async (dispatch) => {
  try {
    const dataURLtoFile = (dataurl, filename) => {
      const arr = dataurl.split(',');
      const mime = arr[0].match(/:(.*?);/)[1];
      const bstr = atob(arr[1]);
      let n = bstr.length;
      const u8arr = new Uint8Array(n);
      while (n) {
        u8arr[n - 1] = bstr.charCodeAt(n - 1);
        n -= 1;
      }
      return new File([u8arr], filename, { type: mime });
    }
    

    const file = dataURLtoFile(newProfilePicture, 'profile-picture');

    const data = new FormData();
    data.append('profile-picture', file, file.name);

    const { data: { data: { imageUrl } } } = await axios({
      url: `${api}/users/${id}/profile/upload`,
      method: 'post',
      data
    });

  dispatch({
    type: UPDATE_PROFILE,
    payload: { imageUrl }
  });
  } catch ({ response: { status } }) {
    switch (status) {
      case 404:
        return 'Not found.';

      case 500:
      case 502:
        return  'Server error.';

      default:
        return 'Unknown error.';
    }
  }
};

export const userBookmarks = (id) => async (dispatch) => {
  try {
    const { data: { userBookmarks } } = await axios.get(`${api}/bookmarks?userId=${id}`);

    const bookmarkedArticlesPromises = userBookmarks.map(async ({ articleUrl }) => {
      const { data: { data } } = await axios.get(articleUrl);
      return data;
    });

    const bookmarks = await Promise.all(bookmarkedArticlesPromises);

    dispatch({
      type: UPDATE_PROFILE,
      payload: { bookmarks }
    });

  } catch (err) {
    return err;
  }
};

export const removeOwnProfile = () => ({
  type: REMOVE_OWN_PROFILE
});
