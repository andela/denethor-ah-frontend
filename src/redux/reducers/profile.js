import { SET_OWN_PROFILE, UPDATE_PROFILE, GET_USER_BOOKMARKS_SUCCESS, REMOVE_OWN_PROFILE} from '../actions/types';

const profileDefaultState = {

};

export default (state = profileDefaultState, action) => {
  switch (action.type) {
    case SET_OWN_PROFILE:
      return action.payload;

    case UPDATE_PROFILE:
      return {...state, ...action.payload};

    case REMOVE_OWN_PROFILE:
      return {};

    case GET_USER_BOOKMARKS_SUCCESS:
      return { ...state, bookmarks: action.payload }

    default:
      return state;
  }
};
