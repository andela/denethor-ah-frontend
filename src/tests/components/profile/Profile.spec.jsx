import React from 'react';
import { shallow } from 'enzyme';
import { Profile } from '../../../components/profile/Profile';
import profile from '../../mock-data/profile';


const updateProfile = jest.fn();
const uploadProfilePicture = jest.fn();

const props = {
  updateProfile,
  uploadProfilePicture,
  history: {}
}
const wrapper = shallow(<Profile {...props} profile={profile.data} />);

test('Should render component correctly', () => {
  wrapper.update();

  expect(wrapper).toMatchSnapshot();
});

test('Shoould update state on input change', () => {
  wrapper.find('input').at(0).simulate('change', { target: { value: 'Omoefe' } });
  expect(wrapper.state('firstname')).toBe('Omoefe');

  wrapper.find('input').at(1).simulate('change', { target: { value: 'Om' } });
  expect(wrapper.state('lastname')).toBe('Om');

  wrapper.find('input').at(2).simulate('change', { target: { value: 'Om' } });
  expect(wrapper.state('username')).toBe('Om');

  wrapper.find('textarea').simulate('change', { target: { value: 'Omoefe' } });
  expect(wrapper.state('bio')).toBe('Omoefe');
});

test('Shoould validate on input blur', () => {
  wrapper.find('input').at(0).simulate('change', { target: { value: 'Om' } });
  expect(wrapper.state('firstname')).toBe('Om');

  wrapper.find('input').at(0).simulate('blur');
  expect(wrapper.state('firstnameError')).toBe('First name must be at least 3 characters');

  wrapper.find('input').at(0).simulate('change', { target: { value: 'Omo efe' } });
  wrapper.find('input').at(0).simulate('blur');
  expect(wrapper.state('firstnameError')).toBe('Names cannot contain spaces');

  wrapper.find('input').at(1).simulate('blur');
  expect(wrapper.state('lastnameError')).toBe('Last name must be at least 3 characters');

  wrapper.find('input').at(1).simulate('change', { target: { value: 'Omo efe' } });
  wrapper.find('input').at(1).simulate('blur');
  expect(wrapper.state('lastnameError')).toBe('Names cannot contain spaces');

  wrapper.find('input').at(2).simulate('blur');
  expect(wrapper.state('usernameError')).toBe('Username must be at least 3 characters');

  wrapper.find('input').at(2).simulate('change', { target: { value: 'Omo efe' } });
  wrapper.find('input').at(2).simulate('blur');
  expect(wrapper.state('usernameError')).toBe('Username cannot contain spaces');

  wrapper.find('textarea').simulate('blur');
  expect(wrapper.state('bioError')).toBe('Bio must be at least 10 characters if present');
});


