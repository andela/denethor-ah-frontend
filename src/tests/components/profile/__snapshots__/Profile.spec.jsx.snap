// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should indicate if user has not been rated 1`] = `
<div>
  <div
    className="profile__banner-section"
  >
    <div
      className="profile__banner-image"
    >
      <img
        alt="banner background"
        src="/assets/img/banner.png"
      />
    </div>
    <div
      className="profile__transparent-overlay"
    />
    <div
      className="profile__banner-content"
    >
      <div
        className="profile__auto-container"
      >
        <div>
          <div
            className="profile__banner-content-title"
          >
            Omoefe Dukuye
            <div
              className="profile--bio"
            >
              <p />
            </div>
          </div>
          <div
            className="profile__avatar-text-group"
          >
            <div
              className="profile__user-profile-picture"
            >
              <div>
                <e
                  backgroundColor="grey"
                  borderStyle={
                    Object {
                      "border": "none",
                    }
                  }
                  closeIconColor="white"
                  cropColor="white"
                  height={280}
                  label=""
                  labelStyle={
                    Object {
                      "background": "url(/assets/img/camera-icon-gray.png)",
                      "bottom": "45%",
                      "cursor": "pointer",
                      "display": "block",
                      "height": "48px",
                      "left": "55%",
                      "opacity": "1",
                      "position": "absolute",
                      "width": "48px",
                      "zIndex": "45",
                    }
                  }
                  lineWidth={4}
                  mimeTypes="image/jpeg,image/png"
                  minCropRadius={30}
                  mobileScaleSpeed={0.5}
                  onBeforeFileLoad={[Function]}
                  onClose={[Function]}
                  onCrop={[Function]}
                  onFileLoad={[Function]}
                  onImageLoad={[Function]}
                  shadingColor="grey"
                  shadingOpacity={0.6}
                  src={null}
                  width={360}
                />
                <div
                  className="profile-username-rate-group"
                >
                  <div
                    className="profile-header-username"
                  >
                    Omoefe94460
                  </div>
                  <div
                    className="profile-rating-container"
                  >
                    <div
                      className="profile-header-rating"
                    >
                      <RatingStars
                        rateNumber={0}
                      />
                    </div>
                    <div
                      className="profile-header-rated-times"
                    >
                      No ratings yet
                    </div>
                  </div>
                </div>
                <div />
              </div>
              <RoundedImage
                alt="profile picture"
                imageSource="https://platform-lookaside.fbsbx.com/platform/profilepic/?asid=111815579944608&height=50&width=50&ext=1555181201&hash=AeTexMaSviYYeH6G"
              />
            </div>
          </div>
        </div>
        <div
          className="profile__stats"
        >
          <h1>
            <span>
              3
            </span>
            <span>
               Articles
            </span>
          </h1>
          <h1>
            <span>
              0
            </span>
            <span>
               Followers
            </span>
          </h1>
          <h1>
            <span>
              0
            </span>
            <span>
               Following
            </span>
          </h1>
        </div>
      </div>
    </div>
  </div>
  <div
    className="profile__article-section"
  >
    <div
      className="flex filter-form"
    >
      <h2
        className="list-section-header"
      >
        Published Articles
      </h2>
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:43:19.726Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="7e379cf3-b23b-4ce7-bac9-b8db087a8ebd"
        key="7e379cf3-b23b-4ce7-bac9-b8db087a8ebd"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="how-to-speak-so-that-people-want-to-listen-everytime"
        title="How to speak so that people want to listen everytime"
        updatedAt="2019-03-15T08:43:19.726Z"
      />
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:46:24.118Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="dc5cb0dc-750a-4b85-bcbb-a8314faa524f"
        key="dc5cb0dc-750a-4b85-bcbb-a8314faa524f"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="how-to-speak-so-that-people-want-to-listen-everytime"
        title="How to speak so that people want to listen everytime"
        updatedAt="2019-03-15T08:46:24.118Z"
      />
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:46:24.615Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="c5c7c34c-c3c8-49fa-abc0-3ebeb9512418"
        key="c5c7c34c-c3c8-49fa-abc0-3ebeb9512418"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="10-taboos-about-you-should-never-share-on-twitter"
        title="10 Taboos About You Should Never Share On Twitter"
        updatedAt="2019-03-15T08:46:24.615Z"
      />
    </div>
  </div>
</div>
`;

exports[`Should render component correctly with published articles 1`] = `
<div>
  <div
    className="profile__banner-section"
  >
    <div
      className="profile__banner-image"
    >
      <img
        alt="banner background"
        src="/assets/img/banner.png"
      />
    </div>
    <div
      className="profile__transparent-overlay"
    />
    <div
      className="profile__banner-content"
    >
      <div
        className="profile__auto-container"
      >
        <div>
          <div
            className="profile__banner-content-title"
          >
            Omoefe Dukuye
            <div
              className="profile--bio"
            >
              <p />
            </div>
          </div>
          <div
            className="profile__avatar-text-group"
          >
            <div
              className="profile__user-profile-picture"
            >
              <div>
                <e
                  backgroundColor="grey"
                  borderStyle={
                    Object {
                      "border": "none",
                    }
                  }
                  closeIconColor="white"
                  cropColor="white"
                  height={280}
                  label=""
                  labelStyle={
                    Object {
                      "background": "url(/assets/img/camera-icon-gray.png)",
                      "bottom": "45%",
                      "cursor": "pointer",
                      "display": "block",
                      "height": "48px",
                      "left": "55%",
                      "opacity": "1",
                      "position": "absolute",
                      "width": "48px",
                      "zIndex": "45",
                    }
                  }
                  lineWidth={4}
                  mimeTypes="image/jpeg,image/png"
                  minCropRadius={30}
                  mobileScaleSpeed={0.5}
                  onBeforeFileLoad={[Function]}
                  onClose={[Function]}
                  onCrop={[Function]}
                  onFileLoad={[Function]}
                  onImageLoad={[Function]}
                  shadingColor="grey"
                  shadingOpacity={0.6}
                  src={null}
                  width={360}
                />
                <div
                  className="profile-username-rate-group"
                >
                  <div
                    className="profile-header-username"
                  >
                    Omoefe94460
                  </div>
                  <div
                    className="profile-rating-container"
                  >
                    <div
                      className="profile-header-rating"
                    >
                      <RatingStars
                        rateNumber={0}
                      />
                    </div>
                    <div
                      className="profile-header-rated-times"
                    >
                      <span
                        className="spinner spinner--ratings"
                      />
                    </div>
                  </div>
                </div>
                <div />
              </div>
              <RoundedImage
                alt="profile picture"
                imageSource="https://platform-lookaside.fbsbx.com/platform/profilepic/?asid=111815579944608&height=50&width=50&ext=1555181201&hash=AeTexMaSviYYeH6G"
              />
            </div>
          </div>
        </div>
        <div
          className="profile__stats"
        >
          <h1>
            <span>
              3
            </span>
            <span>
               Articles
            </span>
          </h1>
          <h1>
            <span>
              0
            </span>
            <span>
               Followers
            </span>
          </h1>
          <h1>
            <span>
              0
            </span>
            <span>
               Following
            </span>
          </h1>
        </div>
      </div>
    </div>
  </div>
  <div
    className="profile__article-section"
  >
    <div
      className="flex filter-form"
    >
      <h2
        className="list-section-header"
      >
        Published Articles
      </h2>
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:43:19.726Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="7e379cf3-b23b-4ce7-bac9-b8db087a8ebd"
        key="7e379cf3-b23b-4ce7-bac9-b8db087a8ebd"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="how-to-speak-so-that-people-want-to-listen-everytime"
        title="How to speak so that people want to listen everytime"
        updatedAt="2019-03-15T08:43:19.726Z"
      />
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:46:24.118Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="dc5cb0dc-750a-4b85-bcbb-a8314faa524f"
        key="dc5cb0dc-750a-4b85-bcbb-a8314faa524f"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="how-to-speak-so-that-people-want-to-listen-everytime"
        title="How to speak so that people want to listen everytime"
        updatedAt="2019-03-15T08:46:24.118Z"
      />
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:46:24.615Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="c5c7c34c-c3c8-49fa-abc0-3ebeb9512418"
        key="c5c7c34c-c3c8-49fa-abc0-3ebeb9512418"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="10-taboos-about-you-should-never-share-on-twitter"
        title="10 Taboos About You Should Never Share On Twitter"
        updatedAt="2019-03-15T08:46:24.615Z"
      />
    </div>
  </div>
</div>
`;

exports[`Should render component correctly without published articles 1`] = `
<div>
  <div
    className="profile__banner-section"
  >
    <div
      className="profile__banner-image"
    >
      <img
        alt="banner background"
        src="/assets/img/banner.png"
      />
    </div>
    <div
      className="profile__transparent-overlay"
    />
    <div
      className="profile__banner-content"
    >
      <div
        className="profile__auto-container"
      >
        <div>
          <div
            className="profile__banner-content-title"
          >
            Omoefe Dukuye
            <div
              className="profile--bio"
            >
              <p />
            </div>
          </div>
          <div
            className="profile__avatar-text-group"
          >
            <div
              className="profile__user-profile-picture"
            >
              <div>
                <e
                  backgroundColor="grey"
                  borderStyle={
                    Object {
                      "border": "none",
                    }
                  }
                  closeIconColor="white"
                  cropColor="white"
                  height={280}
                  label=""
                  labelStyle={
                    Object {
                      "background": "url(/assets/img/camera-icon-gray.png)",
                      "bottom": "45%",
                      "cursor": "pointer",
                      "display": "block",
                      "height": "48px",
                      "left": "55%",
                      "opacity": "1",
                      "position": "absolute",
                      "width": "48px",
                      "zIndex": "45",
                    }
                  }
                  lineWidth={4}
                  mimeTypes="image/jpeg,image/png"
                  minCropRadius={30}
                  mobileScaleSpeed={0.5}
                  onBeforeFileLoad={[Function]}
                  onClose={[Function]}
                  onCrop={[Function]}
                  onFileLoad={[Function]}
                  onImageLoad={[Function]}
                  shadingColor="grey"
                  shadingOpacity={0.6}
                  src={null}
                  width={360}
                />
                <div
                  className="profile-username-rate-group"
                >
                  <div
                    className="profile-header-username"
                  >
                    Omoefe94460
                  </div>
                  <div
                    className="profile-rating-container"
                  >
                    <div
                      className="profile-header-rating"
                    >
                      <RatingStars
                        rateNumber={0}
                      />
                    </div>
                    <div
                      className="profile-header-rated-times"
                    >
                      <span
                        className="spinner spinner--ratings"
                      />
                    </div>
                  </div>
                </div>
                <div />
              </div>
              <RoundedImage
                alt="profile picture"
                imageSource="https://platform-lookaside.fbsbx.com/platform/profilepic/?asid=111815579944608&height=50&width=50&ext=1555181201&hash=AeTexMaSviYYeH6G"
              />
            </div>
          </div>
        </div>
        <div
          className="profile__stats"
        >
          <h1>
            <span>
              0
            </span>
            <span>
               Articles
            </span>
          </h1>
          <h1>
            <span>
              0
            </span>
            <span>
               Followers
            </span>
          </h1>
          <h1>
            <span>
              0
            </span>
            <span>
               Following
            </span>
          </h1>
        </div>
      </div>
    </div>
  </div>
  <div
    className="profile__article-section"
  >
    <div
      className="flex filter-form"
    >
      <div
        className="profile__article-section__no-articles"
      >
        <p>
          No articles yet.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render nothing if information has not loaded 1`] = `
<div>
  <div
    className="profile__banner-section"
  >
    <div
      className="profile__banner-image"
    >
      <img
        alt="banner background"
        src="/assets/img/banner.png"
      />
    </div>
    <div
      className="profile__transparent-overlay"
    />
    <div
      className="profile__banner-content"
    >
      <div
        className="profile__auto-container"
      >
        <div>
          <div
            className="profile__banner-content-title"
          >
            Omoefe Dukuye
            <div
              className="profile--bio"
            >
              <p />
            </div>
          </div>
          <div
            className="profile__avatar-text-group"
          >
            <div
              className="profile__user-profile-picture"
            >
              <div>
                <e
                  backgroundColor="grey"
                  borderStyle={
                    Object {
                      "border": "none",
                    }
                  }
                  closeIconColor="white"
                  cropColor="white"
                  height={280}
                  label=""
                  labelStyle={
                    Object {
                      "background": "url(/assets/img/camera-icon-gray.png)",
                      "bottom": "45%",
                      "cursor": "pointer",
                      "display": "block",
                      "height": "48px",
                      "left": "55%",
                      "opacity": "1",
                      "position": "absolute",
                      "width": "48px",
                      "zIndex": "45",
                    }
                  }
                  lineWidth={4}
                  mimeTypes="image/jpeg,image/png"
                  minCropRadius={30}
                  mobileScaleSpeed={0.5}
                  onBeforeFileLoad={[Function]}
                  onClose={[Function]}
                  onCrop={[Function]}
                  onFileLoad={[Function]}
                  onImageLoad={[Function]}
                  shadingColor="grey"
                  shadingOpacity={0.6}
                  src={null}
                  width={360}
                />
                <div
                  className="profile-username-rate-group"
                >
                  <div
                    className="profile-header-username"
                  >
                    Omoefe94460
                  </div>
                  <div
                    className="profile-rating-container"
                  >
                    <div
                      className="profile-header-rating"
                    >
                      <RatingStars
                        rateNumber={4}
                      />
                    </div>
                    <div
                      className="profile-header-rated-times"
                    >
                      rated by 50 users
                    </div>
                  </div>
                </div>
                <div />
              </div>
              <RoundedImage
                alt="profile picture"
                imageSource="https://platform-lookaside.fbsbx.com/platform/profilepic/?asid=111815579944608&height=50&width=50&ext=1555181201&hash=AeTexMaSviYYeH6G"
              />
            </div>
          </div>
        </div>
        <div
          className="profile__stats"
        >
          <h1>
            <span>
              0
            </span>
            <span>
               Articles
            </span>
          </h1>
          <h1>
            <span />
            <span>
               Followers
            </span>
          </h1>
          <h1>
            <span />
            <span>
               Following
            </span>
          </h1>
        </div>
      </div>
    </div>
  </div>
  <div
    className="profile__article-section"
  >
    <div
      className="flex filter-form"
    >
      <div
        className="profile__article-section__no-articles"
      >
        <p>
          No articles yet.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should use empty spaces for non-existent fields 1`] = `
<div>
  <div
    className="profile__banner-section"
  >
    <div
      className="profile__banner-image"
    >
      <img
        alt="banner background"
        src="/assets/img/banner.png"
      />
    </div>
    <div
      className="profile__transparent-overlay"
    />
    <div
      className="profile__banner-content"
    >
      <div
        className="profile__auto-container"
      >
        <div>
          <div
            className="profile__banner-content-title"
          >
             
            <div
              className="profile--bio"
            >
              <p />
            </div>
          </div>
          <div
            className="profile__avatar-text-group"
          >
            <div
              className="profile__user-profile-picture"
            >
              <div>
                <e
                  backgroundColor="grey"
                  borderStyle={
                    Object {
                      "border": "none",
                    }
                  }
                  closeIconColor="white"
                  cropColor="white"
                  height={280}
                  label=""
                  labelStyle={
                    Object {
                      "background": "url(/assets/img/camera-icon-gray.png)",
                      "bottom": "45%",
                      "cursor": "pointer",
                      "display": "block",
                      "height": "48px",
                      "left": "55%",
                      "opacity": "1",
                      "position": "absolute",
                      "width": "48px",
                      "zIndex": "45",
                    }
                  }
                  lineWidth={4}
                  mimeTypes="image/jpeg,image/png"
                  minCropRadius={30}
                  mobileScaleSpeed={0.5}
                  onBeforeFileLoad={[Function]}
                  onClose={[Function]}
                  onCrop={[Function]}
                  onFileLoad={[Function]}
                  onImageLoad={[Function]}
                  shadingColor="grey"
                  shadingOpacity={0.6}
                  src={null}
                  width={360}
                />
                <div
                  className="profile-username-rate-group"
                >
                  <div
                    className="profile-header-username"
                  />
                  <div
                    className="profile-rating-container"
                  >
                    <div
                      className="profile-header-rating"
                    >
                      <RatingStars
                        rateNumber={0}
                      />
                    </div>
                    <div
                      className="profile-header-rated-times"
                    >
                      <span
                        className="spinner spinner--ratings"
                      />
                    </div>
                  </div>
                </div>
                <div />
              </div>
              <RoundedImage
                alt="profile picture"
                imageSource="https://platform-lookaside.fbsbx.com/platform/profilepic/?asid=111815579944608&height=50&width=50&ext=1555181201&hash=AeTexMaSviYYeH6G"
              />
            </div>
          </div>
        </div>
        <div
          className="profile__stats"
        >
          <h1>
            <span>
              3
            </span>
            <span>
               Articles
            </span>
          </h1>
          <h1>
            <span>
              0
            </span>
            <span>
               Followers
            </span>
          </h1>
          <h1>
            <span>
              0
            </span>
            <span>
               Following
            </span>
          </h1>
        </div>
      </div>
    </div>
  </div>
  <div
    className="profile__article-section"
  >
    <div
      className="flex filter-form"
    >
      <h2
        className="list-section-header"
      >
        Published Articles
      </h2>
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:43:19.726Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="7e379cf3-b23b-4ce7-bac9-b8db087a8ebd"
        key="7e379cf3-b23b-4ce7-bac9-b8db087a8ebd"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="how-to-speak-so-that-people-want-to-listen-everytime"
        title="How to speak so that people want to listen everytime"
        updatedAt="2019-03-15T08:43:19.726Z"
      />
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:46:24.118Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="dc5cb0dc-750a-4b85-bcbb-a8314faa524f"
        key="dc5cb0dc-750a-4b85-bcbb-a8314faa524f"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="how-to-speak-so-that-people-want-to-listen-everytime"
        title="How to speak so that people want to listen everytime"
        updatedAt="2019-03-15T08:46:24.118Z"
      />
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:46:24.615Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="c5c7c34c-c3c8-49fa-abc0-3ebeb9512418"
        key="c5c7c34c-c3c8-49fa-abc0-3ebeb9512418"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="10-taboos-about-you-should-never-share-on-twitter"
        title="10 Taboos About You Should Never Share On Twitter"
        updatedAt="2019-03-15T08:46:24.615Z"
      />
    </div>
  </div>
</div>
`;

exports[`Should use singular if one rater 1`] = `
<div>
  <div
    className="profile__banner-section"
  >
    <div
      className="profile__banner-image"
    >
      <img
        alt="banner background"
        src="/assets/img/banner.png"
      />
    </div>
    <div
      className="profile__transparent-overlay"
    />
    <div
      className="profile__banner-content"
    >
      <div
        className="profile__auto-container"
      >
        <div>
          <div
            className="profile__banner-content-title"
          >
            Omoefe Dukuye
            <div
              className="profile--bio"
            >
              <p />
            </div>
          </div>
          <div
            className="profile__avatar-text-group"
          >
            <div
              className="profile__user-profile-picture"
            >
              <div>
                <e
                  backgroundColor="grey"
                  borderStyle={
                    Object {
                      "border": "none",
                    }
                  }
                  closeIconColor="white"
                  cropColor="white"
                  height={280}
                  label=""
                  labelStyle={
                    Object {
                      "background": "url(/assets/img/camera-icon-gray.png)",
                      "bottom": "45%",
                      "cursor": "pointer",
                      "display": "block",
                      "height": "48px",
                      "left": "55%",
                      "opacity": "1",
                      "position": "absolute",
                      "width": "48px",
                      "zIndex": "45",
                    }
                  }
                  lineWidth={4}
                  mimeTypes="image/jpeg,image/png"
                  minCropRadius={30}
                  mobileScaleSpeed={0.5}
                  onBeforeFileLoad={[Function]}
                  onClose={[Function]}
                  onCrop={[Function]}
                  onFileLoad={[Function]}
                  onImageLoad={[Function]}
                  shadingColor="grey"
                  shadingOpacity={0.6}
                  src={null}
                  width={360}
                />
                <div
                  className="profile-username-rate-group"
                >
                  <div
                    className="profile-header-username"
                  >
                    Omoefe94460
                  </div>
                  <div
                    className="profile-rating-container"
                  >
                    <div
                      className="profile-header-rating"
                    >
                      <RatingStars
                        rateNumber={4}
                      />
                    </div>
                    <div
                      className="profile-header-rated-times"
                    >
                      rated by 1 user
                    </div>
                  </div>
                </div>
                <div />
              </div>
              <RoundedImage
                alt="profile picture"
                imageSource="https://platform-lookaside.fbsbx.com/platform/profilepic/?asid=111815579944608&height=50&width=50&ext=1555181201&hash=AeTexMaSviYYeH6G"
              />
            </div>
          </div>
        </div>
        <div
          className="profile__stats"
        >
          <h1>
            <span>
              3
            </span>
            <span>
               Articles
            </span>
          </h1>
          <h1>
            <span>
              0
            </span>
            <span>
               Followers
            </span>
          </h1>
          <h1>
            <span>
              0
            </span>
            <span>
               Following
            </span>
          </h1>
        </div>
      </div>
    </div>
  </div>
  <div
    className="profile__article-section"
  >
    <div
      className="flex filter-form"
    >
      <h2
        className="list-section-header"
      >
        Published Articles
      </h2>
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:43:19.726Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="7e379cf3-b23b-4ce7-bac9-b8db087a8ebd"
        key="7e379cf3-b23b-4ce7-bac9-b8db087a8ebd"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="how-to-speak-so-that-people-want-to-listen-everytime"
        title="How to speak so that people want to listen everytime"
        updatedAt="2019-03-15T08:43:19.726Z"
      />
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:46:24.118Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="dc5cb0dc-750a-4b85-bcbb-a8314faa524f"
        key="dc5cb0dc-750a-4b85-bcbb-a8314faa524f"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="how-to-speak-so-that-people-want-to-listen-everytime"
        title="How to speak so that people want to listen everytime"
        updatedAt="2019-03-15T08:46:24.118Z"
      />
      <FeedBottomSmallCard
        authorId="2c17e943-df50-485c-870a-1cf3bc59d015"
        body="body"
        categoryId={1}
        createdAt="2019-03-15T08:46:24.615Z"
        dateCreated="17 days ago"
        description="How to speak so that people want to listen everytime"
        featuredImage=""
        id="c5c7c34c-c3c8-49fa-abc0-3ebeb9512418"
        key="c5c7c34c-c3c8-49fa-abc0-3ebeb9512418"
        references={
          Array [
            "reference1.site",
            "reference2.site",
            "reference3.site",
          ]
        }
        slug="10-taboos-about-you-should-never-share-on-twitter"
        title="10 Taboos About You Should Never Share On Twitter"
        updatedAt="2019-03-15T08:46:24.615Z"
      />
    </div>
  </div>
</div>
`;
